src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -83,29 +93,38 @@ object GroupReadsByUmi {
         tmp
       }
       else {
-        val lib       = library(rec)
-        val chrom     = rec.refIndex
-        val mateChrom = rec.mateRefIndex
-        val recNeg    = rec.negativeStrand
-        val recPos    = if (recNeg) rec.unclippedEnd else rec.unclippedStart
-
-        val (mateNeg, matePos) = if (!rec.paired) (false, Int.MaxValue) else {
-          val neg = rec.mateNegativeStrand
-          val pos = if (neg) SAMUtils.getMateUnclippedEnd(rec.asSam) else SAMUtils.getMateUnclippedStart(rec.asSam)
-          (neg, pos)
+        val lib = library(rec)
+
+        val (ref1, start1, strand1) = if (rec.unmapped) (UnknownRef, UnknownPos, UnknownStrand) else {
+          val start = if (rec.positiveStrand) rec.unclippedStart else rec.unclippedEnd
+          (rec.refIndex, start, strandToByte(rec.positiveStrand))
+        }
+
+        val (ref2, start2, strand2) = if (rec.unpaired || rec.mateUnmapped) (UnknownRef, UnknownPos, UnknownStrand) else {
+          val start = if (rec.matePositiveStrand) SAMUtils.getMateUnclippedStart(rec.asSam) else SAMUtils.getMateUnclippedEnd(rec.asSam)
+          (rec.mateRefIndex, start, strandToByte(rec.matePositiveStrand))
         }
 
-        val result = if (chrom < mateChrom || (chrom == mateChrom && (recPos < matePos || (recPos == matePos && !recNeg)))) {
-          new ReadInfo(chrom, recPos, matePos, recNeg, mateNeg, lib)
+        val earlier =
+          (ref1 < ref2) ||
+          (ref1 == ref2 && start1 < start2) ||
+          (ref1 == ref2 && start1 == start2 && strand1 < strand2)
+
+
+        val result = if (earlier) {
+          new ReadInfo(ref1, start1, strand1, ref2, start2, strand2, lib)
         }
         else {
-          new ReadInfo(mateChrom, matePos, recPos, mateNeg, recNeg, lib)
+          new ReadInfo(ref2, start2, strand2, ref1, start1, strand1, lib)
```

Comment:
Since you have duplicate types (many integers) I tend to do by-name parameterization to protect from accidentally swapping arguments.